src/python/pants/backend/python/util_rules/pex_from_targets.py

```diff
@@ -538,7 +539,7 @@ async def create_pex_from_targets(
         sources_digests.append(request.additional_sources)
     if request.include_source_files:
         transitive_targets = await Get(
-            TransitiveTargets, TransitiveTargetsRequest(request.addresses)
+            TransitiveTargets, TransitiveTargetsWithoutTraversingPackagesRequest(request.addresses)
```

Comment:
This is the change required to fix #15855 (this line, and the rest of the logic behind it).